Blog: functional programming constructs (#7)
divyanshu013/blog: content/blog/functional-programming-constructs/index.md

## Why functional programming?

Functional programming aims at making code more readable and understandable. We spend more time reading code than writing it, which justifies the returns to me.

> Code that you cannot trust is code that you do not understand
>
> — [Kyle Simpson](https://twitter.com/getify)

Consider an expression using the `map` keyword. Same logic can be written using a `for` loop, however with the `for` loop we've to read through the entire block to figure out what it does, condition for breaking out of the loop and the values being changed (mutated) inside the `for` loop. When using a `map` a lot of this is implicitly understood by the programmer, therefore less code to read. For example:

```js
// Create a new array from an existing array incrementing each value by 1
// [1, 2, 3] -> [2, 3, 4]

const originalArray = [1, 2, 3];

// First approach using a loop
let newArray = [];
// Note: don't use for...in for iterating over arrays since it returns array indices unordered
for (let value of originalArray) {

> 	newArray.push(value + 1);

Reviewer: fix indentation

Author: The file runs through prettier, so indentation is auto formatted.

### Declarative and imperative

Functional programming is **relatively more declarative**. Imperative code is focused on **how** to do something. Declarative code is focused on describing **what** the code does. If you recall the above example, using a `map` keyword to generate an array of values is **relatively more declarative** than imperatively writing the logic using a `for...of` loop.

### Composition

Composition is combining multiple functions together. Functions can be combined in different ways to achieve different tasks. A very simple example of composition:

```js
// Create a new array from an existing array incrementing each value by 1 and multiplying by 2
// [1, 2, 3] -> [4, 6, 8]

const add1 = num => num + 1;
const mul2 = num => num * 2;

// compose a new function using the above two
// note the order of function invocation matters
const addAndMultiply = num => mul2(add1(num));

// use the composed function
const array = [1, 2, 3].map(num => addAndMultiply(num)); // [4, 6, 8]

// or we can simplify it even more by eliminating the num parameter (make it point-free)
const anotherArray = [1, 2, 3].map(addAndMultiply); // [4, 6, 8]
```

> The biggest benefit of composition is it leads to better abstractions. Composition also leads to clearer code and reusability as compared to classical inheritance based approach. Here's a quote from the [Functional Light JavaScript](https://github.com/getify/Functional-Light-JS) book, which has become my favorite explanation for abstraction:

Reviewer 2: (optional)

```suggestion
The biggest benefit of composition is: it leads to better abstractions. Composition also leads to clearer code and reusability as compared to classical inheritance based approach. Here's a quote from the [Functional Light JavaScript](https://github.com/getify/Functional-Light-JS) book, which has become my favorite explanation for abstraction:
```

> ... abstraction is a process by which the programmer associates a name with a potentially complicated program fragment, which can then be thought of in terms of its purpose of function, rather than in terms of how that function is achieved.
>
> By hiding irrelevant details, abstraction reduces conceptual complexity, making it possible for the programmer to focus on a manageable subset of the program text at any particular time.
>
> — **Michael L. Scott, Programming Language Pragmatics**

### Mutation and side effects

Functional programming avoids mutation of state and side effects. These may include modifying variables outside the scope of function, database or network operations, DOM manipulation and even using random numbers.

#### Mutation

Functional programming favors **value immutability**. Whenever we need to change the state in a program, we should create and track a new value for the state (instead of mutating the existing one).

#### Side effects

> Its impossible to do anything useful in a software without doing any side effects, the existence or non existence of such a software would be equivalent. Instead, functional programming encourages us to make such side effects **more obvious** and **isolated**.

Reviewer 2: ```suggestion
It is impossible to do anything useful in a software without causing any side effects, the existence or non existence of such a software would be equivalent. Instead, functional programming encourages us to make such side effects **more obvious** and **isolated**.
```

> One of the ways of achieving this is by **favoring pure functions over impure**. A pure function doesn't rely on external factors and doesn't cause any side effects. Side effects are instead isolated and handled with impure functions. For example:

Reviewer 2: (optional)

```suggestion
One of the ways of achieving this is by **favoring pure functions over impure**. A pure function neither relies on any external factors nor causes any side effects. Side effects are instead isolated and handled with impure functions. For example:
```
